---
title: "R Notebook"
output:
  html_notebook: default
  html_document:
    df_print: paged
  pdf_document: default
---

```{r, echo=FALSE}
require(data.table)
require(ggplot2)
require(dplyr)
require(tidyr)
require(cowplot)
require(gtable)
require(facetscales)
require(ggpubr)
require(grid)
require(gridExtra)
```



```{r , echo=FALSE, warning=FALSE}
ds = 'mnist-6'
text = 'MNIST'
x = file.path('/home/anne/Documents/featurecloud/singular-value-decomposition/results/', ds)
combine_data<-function(x){
  # Combine vector and matrix data sets.
  data1.0<-fread(file.path(x, 'matrix','summary.angles.u.tsv'))
  colnames(data1.0)<- c('iteration', paste0('Eigenvector ', 1:(ncol(data1.0)-4)), 'sites', 'Algorithm', 'filename')
  data1.0<- data1.0 %>% pivot_longer(-c(iteration, sites, Algorithm, filename), names_to='eigenvector', values_to = 'angle')
  data.1.0.c <- data1.0
  
  # fill the data frame with the last angle for runs that finished early.
  mi.global<-data1.0 %>% group_by(eigenvector, sites, Algorithm)%>% summarise(mi.g = max(iteration))
  mi.file<-data1.0 %>% group_by(eigenvector, sites, Algorithm, filename)%>% summarise(mi.f = max(iteration), mi.a = min(angle))
  mi.file<-merge(mi.file, mi.global, by = c('eigenvector', 'sites', 'Algorithm'))
  mi.file$end_it<-mi.file$mi.g-mi.file$mi.f
  mi.file <- mi.file %>% group_by(eigenvector, sites, Algorithm, filename) %>% slice(rep(1:n(), each = end_it))    
  
  mi.file.s <- mi.file %>%
            group_by(eigenvector, sites,Algorithm, filename) %>%
            mutate(mi.f=row_number(mi.f)+mi.f) %>%
            ungroup() %>% 
            select(mi.f, sites, Algorithm, filename, eigenvector, mi.a)
  colnames(mi.file.s)<-colnames(data1.0)
  data1.0<-rbind(data1.0, mi.file.s)

  
  data1<- data1.0 %>%
    group_by(iteration, sites, Algorithm, eigenvector) %>% summarise(mean_angle= mean(angle))
  
  data1<-as.data.table(data1)
  data1$eigenvector<- sapply(data1$eigenvector, function(x) as.numeric(gsub('Eigenvector', '', x)))
  
  data2.0<-fread(file.path(x, 'vector','summary.angles.u.tsv'))
  colnames(data2.0)<-  c('eigenvector', 'iteration', 'angle' , 'sites', 'Algorithm', 'filename')
  data.2.0.c<-data2.0
  # fill the data frame with the last angle for runs that finished early.
  mi.global<-data2.0 %>% group_by(eigenvector, sites, Algorithm)%>% summarise(mi.g = max(iteration))
  mi.file<-data2.0 %>% group_by(eigenvector, sites, Algorithm, filename)%>% summarise(mi.f = max(iteration), mi.a = min(angle))
  mi.file<-merge(mi.file, mi.global, by = c('eigenvector', 'sites', 'Algorithm'))
  mi.file$end_it<-mi.file$mi.g-mi.file$mi.f
  mi.file <- mi.file %>% group_by(eigenvector, sites, Algorithm, filename) %>% slice(rep(1:n(), each = end_it))

  mi.file.s <- mi.file %>%
            group_by(eigenvector, sites,Algorithm, filename) %>%
            mutate(mi.f.adj=row_number(mi.f)+mi.f) %>%
            ungroup() %>%
            select(eigenvector, mi.f.adj, mi.a, sites, Algorithm, filename, mi.f)
  colnames(mi.file.s)<-c(colnames(data2.0), 'max_it')
  data2.0<-rbind(data2.0, mi.file.s[,1:6])

  data2.0[, index:=1:.N, by=c('sites', 'Algorithm', 'eigenvector', 'filename')]
  data2 <- data2.0 %>% group_by(index, sites, Algorithm, eigenvector) %>% summarise(mean_angle= mean(angle))
  data2<- as.data.table(data2)
  colnames(data2)<-colnames(data1)
  
  
  # Create an offset vector for vizualisation
  offset<-mi.file.s  %>% group_by(eigenvector, sites, Algorithm)%>% summarise(offset=max(max_it)) %>% summarise(mean_offset = floor(mean(offset)))
  offset$eigenvector<-offset$eigenvector+1
  offset$mean_offset<-offset$mean_offset-1
  
  
  #make sure the convergence line starts at the mean iteration to have one line
  data2<- data2 %>% left_join(offset)
  data2$iteration<-rowSums(data2[, c("iteration", "mean_offset")], na.rm = T)
  data2<-data2[, c("iteration", "sites", "Algorithm", "eigenvector", "mean_angle")]
  data2<-as.data.table(data2)
  
  # save a dataframe containing the iterations.
  iterations <-  rbind(data.1.0.c, data.2.0.c[, 1:6, with=F]) %>% group_by(sites, Algorithm, filename) %>% summarise(maxit = max(iteration))
  
  
  # fill the data frame with 90 before the start of gradient convergence
  expanded<-list()
  for (i in 1:(length(offset$eigenvector)-1)){
    expanded[[i]]<-data.table(seq(1,  offset$mean_offset[i]), rep(offset$sites[i], offset$mean_offset[i]), rep('gradient', offset$mean_offset[i]), rep(offset$eigenvector[i], offset$mean_offset[i]), rep(90, offset$mean_offset[i]))
  }
  
  # Combine the data sets
  data<- rbind(data1, data2)
  
  expanded<-rbindlist(expanded)
  data<- rbind(data, expanded, use.names=F)
  
  # add nice facet labels
  data$eigenvector<-paste0('Eigenvector ', data$eigenvector)
  data$eigenvector<-ordered(data$eigenvector, levels = c("Eigenvector 1", "Eigenvector 2",  "Eigenvector 3" , "Eigenvector 4",  "Eigenvector 5" , "Eigenvector 6",  "Eigenvector 7" , "Eigenvector 8",  "Eigenvector 9",  "Eigenvector 10"))
  data$Algorithm<-recode(data$Algorithm, 'power-approx'='AI-FULL', 'power-iteration'='RI-FULL', 'randomized'='RANDOMIZED', 'gradient'='GUO')
  
  return(list(iterations, data))
}

```

## MNIST

```{r, show=F, echo=FALSE, warning=FALSE}
# manually fix the scales for the convergence plots
scales_x <- list(
  `Eigenvector 1` = scale_x_continuous(limits = c(1, 11)),
  `Eigenvector 3` = scale_x_continuous(limits = c(1, 200)),
  `Eigenvector 10` =scale_x_continuous(limits = c(1,220))
  
)

#manually set the breaks for the convergence plots
breaks_fun <- function(x) {
  if (max(x)>1000){
    seq(500, max(x), 500)
  }
  else if (max(x) > 500) {
    seq(250, max(x), 250)
  } else if (max(x)> 250) {
    seq(100, max(x), 150)
  }
  else{
    seq(50, max(x), 100)
  }
}
# make a nice minimal theme
extra_theme <-theme(axis.text.y = element_text( size=12),
                    axis.title = element_text(size=12),
                    plot.margin = unit(c(0.2, 0.2, 0.2, 0.2), "cm"),
                    legend.margin = margin(t = -12, r = 0, b = 0, l = 0),
                    axis.text.x = element_text(angle = 90 , size = 12, hjust = 1, vjust = 0.5, margin = margin(t = 1, r = 0, b = 0, l = 0)))

```

```{r, echo=FALSE, warning=FALSE}
ob<-combine_data(file.path('/home/anne/Documents/featurecloud/singular-value-decomposition/results/',ds))
data<-ob[[2]]
iterations<-ob[[1]]
```


```{r, echo=F}
# plot the convergence lines.
marker<-data[(iteration %% 20 ==0  | eigenvector == 'Eigenvector 1' & iteration %in% c(2,4,6,8)) & eigenvector %in% c('Eigenvector 1',  'Eigenvector 3', 'Eigenvector 10')]
conv<-ggplot(data[eigenvector %in% c('Eigenvector 1',  'Eigenvector 3', 'Eigenvector 10')], 
             aes(iteration, mean_angle, group=interaction(Algorithm, Algorithm), color=Algorithm))+
  geom_line(size=1)+ geom_point(data=marker, aes(x=iteration, y = mean_angle, shape=Algorithm,  group=interaction(Algorithm, Algorithm), color=Algorithm),size=2)+
  xlab('Iterations')+
  theme_classic()+
  ylab('Angles w.r.t reference')+
  facet_grid_sc(rows=vars(sites), cols = vars(eigenvector), scales = list(x = scales_x))+
  scale_x_continuous('Algorithm',breaks=breaks_fun, limits = c(0, NA))+
  theme(legend.position = 'bottom',
        strip.text = element_text(size=12), panel.border = element_rect(fill=NA), 
        panel.spacing = unit(0.05, 'cm'),strip.background = element_blank(), 
        strip.text.y = element_text(size=12))+ # remove sites strip label
  scale_color_viridis_d(option = 'A', end=0.7)+extra_theme


#Make side panel
# Convert the plot to a grob

size=12
gt <- ggplotGrob(conv)
# Get the positions of the right strips in the layout: t = top, l = left, ...
strip <-c(subset(gt$layout, grepl("strip-r", gt$layout$name), select = t:r))
# Text grob
text.grob = textGrob(text, rot = -90,
                     gp = gpar(fontsize = size))
# New olumn to the right of current strip
# Adjusts its width to text size
width = unit(2, "grobwidth", text.grob) + unit(1, "lines")
gt <- gtable_add_cols(gt, width, max(strip$r))  
# Add text grob to new column
gt <- gtable_add_grob(gt, text.grob, 
                      t = min(strip$t), l = max(strip$r) + 1, b = max(strip$b))
# Draw it
conv<-ggplotify::as.ggplot(gt)
```

```{r show=TRUE}
conv
```


```{r , echo=FALSE}

ggsave(conv,file=paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/plots/convergence_',ds,'.pdf'), width = 18, units = 'cm', height = 8)
```
```{r, echo=FALSE, warning=FALSE}
overlay<-data.table(grey = c("AI-FULL", "RANDOMIZED"))
eog<- fread(file.path('~/Documents/featurecloud/singular-value-decomposition/results/',ds,'/matrix/summary.eo.tsv'))
eo<-fread(file.path('/home/anne/Documents/featurecloud/singular-value-decomposition/results/',ds,'/vector/summary.eo.tsv'))
eo<-rbind(eo, eog, use.names=F)
eo$V4<-recode(eo$V4, 'power-approx'='AI-FULL', 'power-iteration'='RI-FULL', 'randomized'='RANDOMIZED', 'gradient'='GUO')
eo<- eo[V4 %in% c('AI-FULL', 'RI-FULL', 'RANDOMIZED', 'GUO') ]
eo <- eo %>% group_by(V3,V4,V5) %>% summarise(sum=sum(V2))
pl.eo<-ggplot(eo)+geom_boxplot(aes(V4, sum, fill=interaction(V4,V3)))+ylab('Runtime[s] (Matrix oper.)')+xlab('')+
    geom_tile(aes(x = grey, y = max(eo$sum), height = Inf, width = 1),data = overlay,  alpha = 0.2)+
  theme_classic()+
  extra_theme+
  guides(fill=FALSE, color=F)+
  scale_fill_manual(values=rep(c(viridisLite::cividis(2, alpha = 0.8)), each=4))




iterations<-as.data.table(iterations)
iterations$Algorithm<-recode(iterations$Algorithm, 'power-approx'='AI-FULL', 'power-iteration'='RI-FULL', 'randomized'='RANDOMIZED', 'gradient'='GUO')
iterations <- iterations[Algorithm %in% c('AI-FULL', 'RI-FULL', 'RANDOMIZED', 'GUO') ]
iterations<-as.data.table(iterations)
pl.iter<-ggplot(iterations)+
  geom_tile(aes(x = grey, y = max(iterations$maxit), height = Inf, width = 1), data = overlay,  alpha = 0.2)+
  geom_boxplot( aes(Algorithm, maxit, fill=interaction(Algorithm, sites)))+
  xlab('')+
  theme_classic()+
  ylab('Iterations to convergence')+
  extra_theme +
  guides(fill="none", color="none")+
  scale_fill_manual(values=rep(c(viridisLite::cividis(2, alpha = 0.8)), each=4))

transmission<-fread(file.path('~/Documents/featurecloud/singular-value-decomposition/results/',ds,'/matrix/summary.transmission.tsv'))
transmission_guo<-fread(file.path('~/Documents/featurecloud/singular-value-decomposition/results/',ds,'/vector/summary.transmission.tsv'))
transmission<-rbind(transmission_guo, transmission)
transmission<-transmission[V2 %in% c("H_local=CS", "H_global=SC")]
colnames(transmission) <- c('index', 'key', 'iteration', 'site', 'size', 'kA', 'sites', 'algorithm', 'filename')
transmission<-transmission[site==1]
volume<-transmission %>% select(size, sites, algorithm, filename)%>% group_by(sites, algorithm, filename)%>% summarise(total_volume=(sum(size)*4)/1000000000)

volume<-as.data.table(volume)
volume$algorithm<-recode(volume$algorithm, 'power-approx'='AI-FULL', 'power-iteration'='RI-FULL', 'randomized'='RANDOMIZED', 'gradient'='GUO')
volume<- volume[algorithm %in% c( 'RI-FULL', 'AI-FULL', 'RANDOMIZED', 'GUO')]


pl.volume<-ggplot(volume)+
    geom_boxplot(aes(algorithm, total_volume, fill=interaction(algorithm, sites)))+
    geom_tile(aes(x = grey, y = max(volume$total_volume), height = Inf, width = 1),data = overlay,  alpha = 0.2)+

  ylab('Total transmitted\n data [GB]')+
  xlab('')+
  theme_classic()+
  extra_theme+
  guides(fill=FALSE, color=F)+
  scale_fill_manual(values=rep(c(viridisLite::cividis(2, alpha = 0.8)), each=4))

############FAKE ALPHA LEGEND
pl.volume.legend<-ggplot(volume)+
    geom_boxplot(aes(as.factor(sites), total_volume, fill=as.factor(sites)))+
  ylab('Total transmitted data [GB]')+
  xlab('')+
  theme_classic()+
  extra_theme+
  scale_fill_manual('Sites', values=rep(c(viridisLite::cividis(2, alpha = 0.8))))

legend_b <- get_legend(
  pl.volume.legend +theme(legend.position = 'bottom',
                          legend.box.margin = margin(t = -0.75, r = 0, b = -0.75, l = 0,'cm'),
                          legend.text = element_text(size = 12),
                          legend.title = element_text(size = 12))
)

size=12
text.grob = textGrob(text, rot = -90,
                     gp = gpar(fontsize = size))
gp<-ggarrange(pl.iter, pl.eo, pl.volume, ncol=3)
gp<-plot_grid(gp, text.grob, rel_widths = c(1,0.05))
gp<-plot_grid(gp, legend_b, ncol=1, rel_heights = c(1,0.1))
```


```{r, warning=FALSE}
# Convert the plot to a grob
gp
```


```{r, echo=FALSE}
ggsave(file=paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/plots/benchmark_',ds,'.pdf'), width = 18, units = 'cm', height = 7.5)
```

## 1000 Genomes

```{r, echo=False}
# manually fix the scales for the convergence plots
scales_x <- list(
  `Eigenvector 1` = scale_x_continuous(limits = c(1, 11)),
  `Eigenvector 3` = scale_x_continuous(limits = c(1, 100)),
  `Eigenvector 10` =scale_x_continuous(limits = c(1, 750))
  
)

#manually set the breaks for the convergence plots
breaks_fun <- function(x) {
  if (max(x)>1000){
    seq(500, max(x), 500)
  }
  else if (max(x) > 500) {
    seq(250, max(x), 250)
  } else if (max(x)> 250) {
    seq(100, max(x), 150)
  }
  else{
    as.integer(seq(50, max(x), 100))
  }
}
# make a nice minimal theme
extra_theme <-theme(axis.text = element_text( size=12),
                    axis.title = element_text(size=12),
                    plot.margin = unit(c(0.2, 0.2, 0.2, 0.2), "cm"),
                    legend.margin = margin(t = -12, r = 0, b = 0, l = 0),
                    axis.text.x = element_text(angle = 90 , size = 12, hjust = 1, vjust = 0.5, margin = margin(t = 1, r = 0, b = 0, l = 0)))

```

```{r, warning=FALSE, echo=FALSE}
chr <- '1'
ob<-combine_data(paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/1000g/chr',chr,'.100000/'))
data<-ob[[2]]
iterations<-ob[[1]]
data$SNPs<-'100000'
iterations$SNPs<-'100000'

ob1<-combine_data(paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/1000g/chr',chr,'.500000/'))
data1<-ob[[2]]
data1$SNPs<-'500000'
iterations1<-ob1[[1]]
iterations1$SNPs<-'500000'

ob2<-combine_data(paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/1000g/chr',chr,'.all/'))
data2<-ob[[2]]
data2$SNPs<-'>1Mio'
iterations2<-ob2[[1]]
iterations2$SNPs<-'>1Mio'

data<-rbind(data,data1,data2)
iterations<-rbind(iterations, iterations1, iterations2)
iterations$SNPs<-factor(iterations$SNPs, levels = c('100000', '500000', '>1Mio'))
data$SNPs<-factor(data$SNPs, levels = c('100000', '500000', '>1Mio'))

```


```{R, message=FALSE, echo=FALSE}
marker<-data[(eigenvector == 'Eigenvector 3' & iteration %% 10 == 0  | (eigenvector == 'Eigenvector 1' & iteration %in% c(2,4,6,8)) | eigenvector == 'Eigenvector 10'  & iteration %% 100 == 0) & eigenvector %in% c('Eigenvector 1',  'Eigenvector 3', 'Eigenvector 10')]
# plot the convergence lines.
conv<-ggplot(data[eigenvector %in% c('Eigenvector 1', 'Eigenvector 3', 'Eigenvector 10')], 
             aes(iteration, mean_angle, group=interaction(Algorithm, Algorithm), color=Algorithm))+
  geom_line(size=1)+
  geom_point(data=marker, aes(x=iteration, y = mean_angle, shape=Algorithm,  group=interaction(Algorithm, Algorithm), color=Algorithm),size=2)+
  xlab('Iterations')+
  theme_classic()+
  ylab('Angles w.r.t reference')+
  facet_grid_sc(rows=vars(SNPs), cols = vars(eigenvector), scales = list(x = scales_x))+
  scale_x_continuous('Algorithm',breaks=breaks_fun, limits = c(0, NA))+
  theme(legend.position = 'bottom',
        strip.text = element_text(size=12), panel.border = element_rect(fill=NA), 
        panel.spacing = unit(0.05, 'cm'),strip.background = element_blank(),
        axis.title.x = element_text(size=12), axis.title.y = element_text(size = 12))+ # remove sites strip label
  scale_color_viridis_d(option = 'A', end=0.7)+extra_theme

#Make side panel
# Convert the plot to a grob
text = paste0('Chromosome ', chr)
size=12
gt <- ggplotGrob(conv)
# Get the positions of the right strips in the layout: t = top, l = left, ...
strip <-c(subset(gt$layout, grepl("strip-r", gt$layout$name), select = t:r))
# Text grob
text.grob = textGrob(text, rot = -90,
                     gp = gpar(fontsize = size))
# New olumn to the right of current strip
# Adjusts its width to text size
width = unit(2, "grobwidth", text.grob) + unit(1, "lines")
gt <- gtable_add_cols(gt, width, max(strip$r))  
# Add text grob to new column
gt <- gtable_add_grob(gt, text.grob, 
                      t = min(strip$t), l = max(strip$r) + 1, b = max(strip$b))
# Draw it
conv<-ggplotify::as.ggplot(gt)
```

```{r}
conv
```


```{r, echo=FALSE}
ggsave(file=paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/plots/convergence_chromosome_',chr,'.pdf'), width = 18, units = 'cm', height = 12)

```

```{r, echo=FALSE}
eol<-list()

for (sub in c(paste0('chr',chr,'.100000'), paste0('chr',chr,'.500000'), paste0('chr',chr,'.all'))){
  eog<- fread(paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/1000g/',sub,'/matrix/summary.eo.tsv'))
  eo<-fread(paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/1000g/',sub,'/vector/summary.eo.tsv'))
  eo<-rbind(eo, eog, use.names=F)
  eo$SNPs<-as.factor(strsplit(sub, '\\.')[[1]][[2]])
  
  eo$V4<-recode(eo$V4, 'power-approx'='AI-FULL', 'power-iteration'='RI-FULL', 'randomized'='RANDOMIZED', 'gradient'='GUO')
  eo<- eo[V4 %in% c('AI-FULL', 'RI-FULL', 'RANDOMIZED', 'GUO') ]
  eo <- eo %>% group_by(V3,V4,V5, SNPs) %>% summarise(sum=sum(V2))
  eol[[strsplit(sub, '\\.')[[1]][[2]]]]<-eo
}
eo<-rbindlist(eol)

overlay<-data.table(grey = c("AI-FULL", "RANDOMIZED"))
pl.eo<-ggplot(eo)+geom_boxplot( aes(V4, sum, fill=interaction(V4,SNPs)))+
  ylab('Runtime[s] (Matrix oper.)')+xlab('')+
  theme_classic()+
  extra_theme+
  guides(fill=FALSE, color=F)+
  scale_fill_manual(values=rep(c(viridisLite::cividis(3, alpha = 0.8)), each=4))+
  geom_tile(aes(x = grey, y = max(volume$total_volume), height = Inf, width = 1),data = overlay,  alpha = 0.2)

```


```{r, echo=FALSE}
iterations<-as.data.table(iterations)
iterations$Algorithm<-recode(iterations$Algorithm, 'power-approx'='AI-FULL', 'power-iteration'='RI-FULL', 'randomized'='RANDOMIZED', 'gradient'='GUO')
iterations <- iterations[Algorithm %in% c('AI-FULL', 'RI-FULL', 'RANDOMIZED', 'GUO') ]
pl.iter<-ggplot(iterations)+
  geom_boxplot(aes(Algorithm, maxit, fill=interaction(Algorithm, SNPs)))+
  xlab('')+
  theme_classic()+
  ylab('Iterations to convergence')+
  extra_theme +
  guides(fill="none", color="none")+
  scale_fill_manual(values=rep(c(viridisLite::cividis(3, alpha = 0.8)), each=4))+
  geom_tile(aes(x = grey, y = max(volume$total_volume), height = Inf, width = 1),data = overlay,  alpha = 0.2)

```


```{r, echo=FALSE}
tol<-list()
for (sub in c(paste0('chr',chr,'.100000'), paste0('chr',chr,'.500000'), paste0('chr',chr,'.all'))){
  transmission<-fread(paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/1000g/',sub,'/matrix/summary.transmission.tsv'))
transmission_guo<-fread(paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/1000g/',sub,'/vector/summary.transmission.tsv'))
transmission<-rbind(transmission_guo, transmission)
transmission<-transmission[V2 %in% c("H_local=CS", "H_global=SC")]
colnames(transmission) <- c('index', 'key', 'iteration', 'site', 'size', 'kA', 'sites', 'algorithm', 'filename')
transmission<-transmission[site==1]
transmission$SNPs<-strsplit(sub, '\\.')[[1]][[2]]
tol[[strsplit(sub, '\\.')[[1]][[2]]]]<-transmission
}
transmission<-rbindlist(tol)

volume<-transmission %>% select(size, SNPs, algorithm, filename)%>% group_by(SNPs, algorithm, filename)%>% summarise(total_volume=(sum(size)*4)/1000000000)

volume<-as.data.table(volume)
volume$algorithm<-recode(volume$algorithm, 'power-approx'='AI-FULL', 'power-iteration'='RI-FULL', 'randomized'='RANDOMIZED', 'gradient'='GUO')
volume<- volume[algorithm %in% c( 'RI-FULL', 'AI-FULL', 'RANDOMIZED', 'GUO')]
pl.volume<-ggplot(volume)+geom_boxplot(aes(algorithm, total_volume, fill=interaction(algorithm, SNPs)))+
  ylab('Total transmitted data [GB]')+
  xlab('')+
  theme_classic()+
  extra_theme+
  guides(fill=FALSE, color=F)+
    scale_fill_manual(values=rep(c(viridisLite::cividis(3, alpha = 0.8)), each=4))+
  geom_tile(aes(x = grey, y = max(volume$total_volume), height = Inf, width = 1),data = overlay,  alpha = 0.2)

pl.volume
```


```{r, echo=FALSE}
############FAKE ALPHA LEGEND
pl.volume.legend<-ggplot(iterations, aes(SNPs, maxit, fill=SNPs))+
  geom_boxplot()+
  theme_classic()+ylab('Total transmitted data [GB]')+xlab('')+
  scale_fill_manual(values=rep(c(viridisLite::cividis(3, alpha = 0.8))))+
  # Use this to create a legens for splits category
  #guides(fill=F,alpha=guide_legend('#Sites', override.aes=list(fill=hcl(c(15,195),100,0,alpha=c(0.2,0.7)))))
  guides(fill=guide_legend('#SNPs'))
  

legend_b <- get_legend(
  pl.volume.legend +theme(legend.position = 'bottom',
                          legend.box.margin = margin(t = -0.75, r = 0, b = -0.75, l = 0,'cm'),
                          legend.text = element_text(size = 12),
                          legend.title = element_text(size = 12))
)
text = paste0('Chromosome ', chr)
size=12
text.grob = textGrob(text, rot = -90,
                     gp = gpar(fontsize = size))
gp<-ggarrange(pl.iter, pl.eo, pl.volume, ncol=3)
gp<-plot_grid(gp, rel_widths = c(1,0.05))
gp<-plot_grid(gp, legend_b, ncol=1, rel_heights = c(1,0.1))
```


```{r, show=TRUE}
gp
```


```{r, echo=FALSE}
ggsave(gp, file=paste0('/home/anne/Documents/featurecloud/singular-value-decomposition/results/plots/benchmark_chromosome_',chr,'.pdf'), width = 25, units = 'cm', height = 11)
```

